common/auto-updater/http-downloader.cpp

```diff
@@ -0,0 +1,186 @@
+// License: Apache 2.0. See LICENSE file in root directory.
+// Copyright(c) 2020 Intel Corporation. All Rights Reserved.
+
+
+
+#ifdef ENABLE_RS_AUTO_UPDATER
+#include <sstream>
+#include <iostream>
+#include <thread>
+#include <atomic>
+#include <mutex>
+#include <curl/curl.h>
+#include <curl/easy.h>
+#include "types.h"
+#endif // ENABLE_RS_AUTO_UPDATER
+
+#include "http-downloader.h"
+
+
+namespace rs2
+{
+#ifndef ENABLE_RS_AUTO_UPDATER
+    // Dummy functions
+    http_downloader::http_downloader() {};
+    http_downloader::~http_downloader() {};
+    bool http_downloader::download_to_stream(const std::string& url, std::stringstream &output, user_callback_func_type user_callback_func) { return false; }
+    bool http_downloader::download_to_file(const std::string& url, const std::string &file_name, user_callback_func_type user_callback_func) { return false; }
+    bool http_downloader::download_to_bytes_vector(const std::string& url, std::vector<uint8_t> &output, user_callback_func_type user_callback_func) { return false; }
+
+#else
+
+    std::mutex initialize_mutex;
+    static const curl_off_t HALF_SEC = 500000; // User call back function delay
+    static const int CONNECT_TIMEOUT = 5L; // Libcurl download timeout 5 [Sec]
+
+    struct progress_data {
+        curl_off_t last_run_time;
+        user_callback_func_type user_callback_func;
+        CURL *curl;
+    };
+
+
+    size_t stream_write_callback(void *ptr, size_t size, size_t nmemb, void *stream) {
+        std::string data((const char*)ptr, (size_t)size * nmemb);
+        *((std::stringstream*)stream) << data;
+        return size * nmemb;
+    }
+
+    size_t bytes_write_callback(void *ptr, size_t size, size_t nmemb, void *vec) {
+        uint8_t* source_bytes(static_cast<uint8_t*>(ptr));
+
+        int total_size((int)(size * nmemb));
+        while (total_size > 0)
+        {
+            static_cast<std::vector<uint8_t> *>(vec)->push_back(*source_bytes);
+            source_bytes++;
+            --total_size;
+        }
+        
+        return size * nmemb;
+    }
+
+    // This function will be called if CURLOPT_NOPROGRESS is set to 0
+    // Return value: 0 = continue download / 1 = stop download
+    int progress_callback(void *p, curl_off_t dltotal, curl_off_t dlnow, curl_off_t ultotal, curl_off_t ulnow)
```

Comment:
ultotal is not used in the code - is it required?

Reply:
progress_callback is the function libcurl calls and can't be modified.
Since we are using it as download only in the http-downloader I droped it within the callback wrapper function.